In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler, OneHotEncoder
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score,  mean_absolute_percentage_error, root_mean_squared_error
from sklearn.linear_model import LinearRegression, Lasso, Ridge, ElasticNet
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn import impute
from sklearn import preprocessing
import numpy as np

import matplotlib.pyplot as plt
from sklearn.model_selection import cross_val_score, KFold, train_test_split
from sklearn.decomposition import PCA

In [2]:
file_path = r"https://storage.yandexcloud.net/hse-ai24-team-22-data/team-86/data/resultwithfeatures.csv?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=YCAJEWD0UOGcR6RTlQ4eOHfGj%2F20250310%2Fru-central1%2Fs3%2Faws4_request&X-Amz-Date=20250310T173632Z&X-Amz-Expires=3600&X-Amz-Signature=91CCB78B1EEF4413AB5D5FF47ABD76CA4EF5347F46725592B29FB98CD73E3698&X-Amz-SignedHeaders=host"
df = pd.read_csv(file_path)

<ipython-input-2-bb0ba0e58f2a>:2: DtypeWarning: Columns (21) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(file_path)


In [3]:
# Предобработка даннных

df = df.drop_duplicates(subset = ['Smiles', 'Exp. Animal', 'Method of administration'])

df = df.drop(['Source', 'Smiles',
       'LD50 (a.u.)', 'Carcinogenicity', 'Carcinogenicity (a.u.)',
       'Hepatoxicity', 'Hepatoxicity (a.u.)', 'Eye_Corrosion',
       'Eye_Corrosion (a.u.)', 'Eye_Irritation', 'Eye_Irritation (a.u.)',
       'Mutagenicity', 'Mutagenicity (a.u.)', 'Respiratory_Toxicity',
       'Respiratory_Toxicity (a.u.)', 'LC50', 'LC50 (a.u.)', 'NOAEL',
       'NOAEL (a.u)'], axis = 1).reset_index(drop = True)

df = df.dropna(subset=['LD50'])

Q10 = df['LD50'].quantile(0.10)
Q90 = df['LD50'].quantile(0.90)
IQR = Q90 - Q10

lower_bound = Q10 - 1.5 * IQR
upper_bound = Q90 + 1.5 * IQR

df = df[(df['LD50'] >= lower_bound) & (df['LD50'] <= upper_bound)]
df = df[df['LD50'] != 0]

threshold = 0.996  # Порог в 99.6%
#удаляем столбцы, где 99.6% значений — это одно уникальное значение
def drop_constant_columns(df, threshold):
    to_drop = []  # Список столбцов для удаления
    for col in df.columns:
        # Рассчитываем долю самого частого значения
        max_freq = df[col].value_counts(normalize=True).max()
        if max_freq >= threshold:
            to_drop.append(col)
    # Удаляем столбцы
    return df.drop(columns=to_drop)

# Применяем функцию
df = drop_constant_columns(df, threshold)
df = df.dropna()


In [10]:
# Разделим данные на тест и треин

df['log_LD50'] = np.log(df['LD50'])
X = df.drop(columns=['LD50', 'log_LD50'])
y = df['log_LD50']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, shuffle=True, random_state=123)

In [11]:
# Кодировка и масштабирование признаков

num_cols = X.select_dtypes(include=['number']).columns.tolist()
cat_cols = X.select_dtypes(include=['object']).columns.tolist()

normalizer = MinMaxScaler()
X_scaled_train = normalizer.fit_transform(X_train[num_cols])
X_scaled_train = pd.DataFrame(data=X_scaled_train, columns=num_cols)

X_scaled_test = normalizer.transform(X_test[num_cols])
X_scaled_test = pd.DataFrame(data=X_scaled_test, columns=num_cols)

OHE = OneHotEncoder(handle_unknown='ignore', drop='first')
X_encoded_train = OHE.fit_transform(X_train[cat_cols])
X_encoded_train = pd.DataFrame(X_encoded_train.toarray(), columns=OHE.get_feature_names_out(cat_cols))

X_encoded_test = OHE.transform(X_test[cat_cols])
X_encoded_test = pd.DataFrame(X_encoded_test.toarray(), columns=OHE.get_feature_names_out(cat_cols))

X_train = pd.concat([X_scaled_train, X_encoded_train], axis=1)
X_test = pd.concat([X_scaled_test, X_encoded_test], axis=1)

In [12]:
# Уменьшение размерности с помощью PCA
pca = PCA(n_components=0.95)  # 95% объяснённой дисперсии

# Подгоняем PCA на тренировочных данных и трансформируем их
X_train = pca.fit_transform(X_train)

# Трансформируем тестовые данные с помощью уже подогнанного PCA
X_test = pca.transform(X_test)

In [ ]:
# Построим линейную модель и посмотрим на результаты

model = LinearRegression()
model.fit(X_train, y_train)
preds = model.predict(X_test)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(model.predict(X_train)))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(model.predict(X_test)))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), np.exp(model.predict(X_train)))}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(model.predict(X_test)))}")

MAPE на трейне =22.409831431516203
MAPE на тесте =41.562307209801226
r2 на трейне =0.13413500542090984
r2 на тесте =0.1170494325258673


In [ ]:
# Проверим качество на кросс-валидации по метрике MAPE

cross_val_score(LinearRegression(), X_train, np.exp(y_train), cv=3, scoring='neg_mean_absolute_percentage_error').mean()

-253.82936371218008

In [16]:
cross_val_score(LinearRegression(), X_train, y_train, cv=3, scoring='r2').mean()

0.4275808366551976

In [17]:
cross_val_score(XGBRegressor(), X_train, y_train, cv=3, scoring='r2').mean()

0.525158325465685

In [ ]:
import seaborn as sns

from mlxtend.plotting import plot_decision_regions
from mlxtend.evaluate import bias_variance_decomp
from graphviz import Source
from sklearn.tree import export_graphviz

from sklearn.metrics import accuracy_score, roc_auc_score, RocCurveDisplay
from sklearn.calibration import CalibrationDisplay
from sklearn.isotonic import IsotonicRegression

from sklearn.tree import DecisionTreeRegressor, plot_tree
from sklearn.ensemble import RandomForestRegressor

In [ ]:
model = DecisionTreeRegressor(max_depth=20)
model.fit(X_train, y_train)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(model.predict(X_train)))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(model.predict(X_test)))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), np.exp(model.predict(X_train)))}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(model.predict(X_test)))}")

MAPE на трейне =1.1568583890524515
MAPE на тесте =128.42205875705474
r2 на трейне =0.6529618844078585
r2 на тесте =-0.041469204613793664


In [ ]:
params = {'max_depth' : np.arange(2, 12) }

gs = GridSearchCV(DecisionTreeRegressor(), params, cv=3, scoring='r2', verbose=3)

gs.fit(X_train, y_train)

Fitting 3 folds for each of 10 candidates, totalling 30 fits
[CV 1/3] END .......................max_depth=2;, score=0.060 total time=  15.5s
[CV 2/3] END .......................max_depth=2;, score=0.064 total time=  15.7s
[CV 3/3] END .......................max_depth=2;, score=0.068 total time=  15.2s
[CV 1/3] END .......................max_depth=3;, score=0.087 total time=  22.6s
[CV 2/3] END .......................max_depth=3;, score=0.091 total time=  22.5s
[CV 3/3] END .......................max_depth=3;, score=0.096 total time=  25.7s
[CV 1/3] END .......................max_depth=4;, score=0.124 total time=  29.4s
[CV 2/3] END .......................max_depth=4;, score=0.126 total time=  28.9s
[CV 3/3] END .......................max_depth=4;, score=0.134 total time=  29.1s
[CV 1/3] END .......................max_depth=5;, score=0.141 total time=  35.9s
[CV 2/3] END .......................max_depth=5;, score=0.149 total time=  35.8s
[CV 3/3] END .......................max_depth=5;

GridSearchCV(cv=3, estimator=DecisionTreeRegressor(),
             param_grid={'max_depth': array([ 2,  3,  4,  5,  6,  7,  8,  9, 10, 11])},
             scoring='r2', verbose=3)

In [ ]:
gs.best_estimator_, gs.best_score_

(DecisionTreeRegressor(max_depth=8), 0.1889312178919159)

In [ ]:
dt = gs.best_estimator_

train_preds = dt.predict(X_train)
test_preds = dt.predict(X_test)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(train_preds))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(test_preds))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), train_preds)}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(test_preds))}")

MAPE на трейне =29.821630296979553
MAPE на тесте =53.63507631389726
r2 на трейне =-0.47390495728377147
r2 на тесте =-0.043958958300331474


In [ ]:
params = {'max_depth' : [11, 12],
          'max_features' : [None, "sqrt", "log2"]}

gs = GridSearchCV(DecisionTreeRegressor(), params, cv=3, scoring='r2', verbose=3)

gs.fit(X_train, y_train)

Fitting 3 folds for each of 6 candidates, totalling 18 fits
[CV 1/3] END ...max_depth=11, max_features=None;, score=0.169 total time= 1.2min
[CV 2/3] END ...max_depth=11, max_features=None;, score=0.160 total time= 1.2min
[CV 3/3] END ...max_depth=11, max_features=None;, score=0.160 total time= 1.2min
[CV 1/3] END ...max_depth=11, max_features=sqrt;, score=0.072 total time=   3.4s
[CV 2/3] END ...max_depth=11, max_features=sqrt;, score=0.089 total time=   2.8s
[CV 3/3] END ...max_depth=11, max_features=sqrt;, score=0.098 total time=   3.2s
[CV 1/3] END ...max_depth=11, max_features=log2;, score=0.022 total time=   1.5s
[CV 2/3] END ...max_depth=11, max_features=log2;, score=0.025 total time=   1.2s
[CV 3/3] END ...max_depth=11, max_features=log2;, score=0.052 total time=   1.2s
[CV 1/3] END ...max_depth=12, max_features=None;, score=0.152 total time= 1.2min
[CV 2/3] END ...max_depth=12, max_features=None;, score=0.150 total time= 1.3min
[CV 3/3] END ...max_depth=12, max_features=None;,

GridSearchCV(cv=3, estimator=DecisionTreeRegressor(),
             param_grid={'max_depth': [11, 12],
                         'max_features': [None, 'sqrt', 'log2']},
             scoring='r2', verbose=3)

In [ ]:
gs.best_estimator_, gs.best_score_

(DecisionTreeRegressor(max_depth=11), 0.16309247773985158)

In [ ]:
dt = gs.best_estimator_

train_preds = dt.predict(X_train)
test_preds = dt.predict(X_test)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(train_preds))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(test_preds))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), np.exp(train_preds))}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(test_preds))}")

MAPE на трейне =16.25415926824466
MAPE на тесте =49.360855940656016
r2 на трейне =0.1202027145257325
r2 на тесте =0.023259471421516076


In [ ]:
rf = RandomForestRegressor(random_state = 123, n_estimators=5, max_depth=11)
rf.fit(X_train, y_train)

train_preds = rf.predict(X_train)
test_preds = rf.predict(X_test)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(train_preds))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(test_preds))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), np.exp(train_preds))}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(test_preds))}")

MAPE на трейне =13.512880844939303
MAPE на тесте =44.718273501877356
r2 на трейне =0.05193595102391679
r2 на тесте =-0.0057825433164413464


In [ ]:
# Определяем сетку гиперпараметров
param_dist = {
    "n_estimators": np.arange(50, 301, 50),
    "max_depth": [None, 10, 20, 30, 50],
    "min_samples_leaf": [1, 2, 4],
}

# Создаем модель
rf = RandomForestRegressor(random_state=42)

# Запускаем RandomizedSearchCV
random_search = RandomizedSearchCV(
    estimator=rf,
    param_distributions=param_dist,
    n_iter=10,  # Количество итераций
    cv=3,  # 5-фолд кросс-валидация
    scoring="neg_mean_absolute_percentage_error",  # Метрика качества
    n_jobs=-1,  # Используем все ядра процессора
    verbose=3
)

# Обучаем модель
random_search.fit(X_train, y_train)

# Выводим лучшие параметры
print("Лучшие параметры случайного леса:", random_search.best_params_)

Fitting 3 folds for each of 10 candidates, totalling 30 fits


In [ ]:
from xgboost import XGBRegressor

boost = XGBRegressor()
boost.fit(X_train, y_train)

train_preds = boost.predict(X_train)
test_preds = boost.predict(X_test)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(train_preds))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(test_preds))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), np.exp(train_preds))}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(test_preds))}")

MAPE на трейне =0.9070029914057807
MAPE на тесте =2.298873468404665
r2 на трейне =0.4977876885936501
r2 на тесте =0.2711146253672051


In [ ]:
!pip install catboost
!pip install lightgbm

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 98.7/98.7 MB 7.0 MB/s eta 0:00:00


In [ ]:
from catboost import CatBoostRegressor

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, shuffle=True, random_state=123)

cat_model = CatBoostRegressor(cat_features=cat_cols, silent=True)
cat_model.fit(X_train, y_train)

train_preds = cat_model.predict(X_train)
test_preds = cat_model.predict(X_test)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(train_preds))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(test_preds))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), np.exp(train_preds))}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(test_preds))}")

MAPE на трейне =2.3601218001600546
MAPE на тесте =4.315919766835618
r2 на трейне =0.4034022730054527
r2 на тесте =0.33934997447637794


In [ ]:
from lightgbm import LGBMRegressor

lightgbm = LGBMRegressor()
lightgbm.fit(X_train, y_train)

train_preds = lightgbm.predict(X_train)
test_preds = lightgbm.predict(X_test)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(train_preds))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(test_preds))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), np.exp(train_preds))}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(test_preds))}")

/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 0.929568 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899


/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


MAPE на трейне =5.547861329508594
MAPE на тесте =19.687955097050185
r2 на трейне =0.16012371924245838
r2 на тесте =0.11495855084776885


In [ ]:
!pip install optuna

In [ ]:
from lightgbm import LGBMRegressor
import optuna

def objective_lgbm(trial):
    max_depth = trial.suggest_int("max_depth", 2, 20)
    learning_rate = trial.suggest_float("learning_rate", 1e-5, 1, log=True)
    n_estimators = trial.suggest_int("n_estimators", 100, 1000)

    score = cross_val_score(LGBMRegressor(max_depth=max_depth, learning_rate=learning_rate, n_estimators=n_estimators),
                            X_train, y_train, cv=3, scoring='neg_mean_absolute_percentage_error', n_jobs=-1).mean()
    return score


study = optuna.create_study(direction="maximize")
study.optimize(objective_lgbm, n_trials=10)

[I 2025-03-09 06:42:48,290] A new study created in memory with name: no-name-5c7d7143-a79f-40ae-8ccf-7d089fe369bb
[I 2025-03-09 06:56:07,518] Trial 0 finished with value: -14547462454660.312 and parameters: {'max_depth': 8, 'learning_rate': 0.00017314070085509657, 'n_estimators': 763}. Best is trial 0 with value: -14547462454660.312.
[I 2025-03-09 07:00:52,454] Trial 1 finished with value: -11661007382207.53 and parameters: {'max_depth': 18, 'learning_rate': 0.054686833377811966, 'n_estimators': 249}. Best is trial 1 with value: -11661007382207.53.
[I 2025-03-09 07:10:26,588] Trial 2 finished with value: -10766218894480.217 and parameters: {'max_depth': 16, 'learning_rate': 0.10385498382901673, 'n_estimators': 623}. Best is trial 2 with value: -10766218894480.217.
[I 2025-03-09 07:13:33,394] Trial 3 finished with value: -14280376267042.078 and parameters: {'max_depth': 19, 'learning_rate': 0.004284594100509765, 'n_estimators': 132}. Best is trial 2 with value: -10766218894480.217.
[I 2

In [ ]:
study.best_params

{'max_depth': 16, 'learning_rate': 0.10385498382901673, 'n_estimators': 623}

In [ ]:
lgbm_model = LGBMRegressor(**study.best_params)
lgbm_model.fit(X_train, y_train)

train_preds = lgbm_model.predict(X_train)
test_preds = lgbm_model.predict(X_test)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(train_preds))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(test_preds))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), np.exp(train_preds))}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(test_preds))}")

/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 1.489380 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899


/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


MAPE на трейне =0.7927336058140227
MAPE на тесте =5.83013318938401
r2 на трейне =0.543391483586178
r2 на тесте =0.28789271804969485


In [ ]:
print(f"MAPE на трейне ={mean_absolute_percentage_error(y_train, train_preds)}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(y_test, test_preds)}")

print(f"r2 на трейне ={r2_score(y_train, train_preds)}")
print(f"r2 на тесте ={r2_score(y_test, test_preds)}")

MAPE на трейне =4670433696601.417
MAPE на тесте =8300822546011.142
r2 на трейне =0.8517425989008165
r2 на тесте =0.6143021293447632


In [ ]:
from lightgbm import LGBMRegressor

def objective_lgbm(trial):
    max_depth = trial.suggest_int("max_depth", 2, 20)
    learning_rate = trial.suggest_float("learning_rate", 1e-5, 1, log=True)
    n_estimators = trial.suggest_int("n_estimators", 100, 1000)

    boost = LGBMRegressor(max_depth=max_depth, learning_rate=learning_rate, n_estimators=n_estimators)
    boost.fit(X_train, y_train)

    test_preds = boost.predict(X_test)

    score = r2_score(y_test, test_preds)
    return score


study = optuna.create_study(direction="maximize")
study.optimize(objective_lgbm, n_trials=10)

[I 2025-03-09 05:32:15,921] A new study created in memory with name: no-name-9ba9182f-5386-4476-888f-eb7a0558a905
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 1.430306 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899


/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
[I 2025-03-09 05:36:37,290] Trial 0 finished with value: 0.5885272665136787 and parameters: {'max_depth': 15, 'learning_rate': 0.060611653131808045, 'n_estimators': 570}. Best is trial 0 with value: 0.5885272665136787.
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 1.425711 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899


/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
[I 2025-03-09 05:40:05,166] Trial 1 finished with value: 0.10339739088951683 and parameters: {'max_depth': 20, 'learning_rate': 0.0010754388318739891, 'n_estimators': 382}. Best is trial 0 with value: 0.5885272665136787.
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 1.413976 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive ga

/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
[I 2025-03-09 05:42:14,607] Trial 2 finished with value: 0.016387420464769265 and parameters: {'max_depth': 4, 'learning_rate': 0.00014871570832738622, 'n_estimators': 472}. Best is trial 0 with value: 0.5885272665136787.
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 1.702278 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899


/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
[I 2025-03-09 05:45:48,526] Trial 3 finished with value: 0.013983698526444766 and parameters: {'max_depth': 9, 'learning_rate': 0.0001070918652184647, 'n_estimators': 413}. Best is trial 0 with value: 0.5885272665136787.
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 1.355712 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899


/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
[I 2025-03-09 05:48:47,715] Trial 4 finished with value: 0.01837354616622844 and parameters: {'max_depth': 14, 'learning_rate': 0.00017593998805313273, 'n_estimators': 332}. Best is trial 0 with value: 0.5885272665136787.
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 1.409565 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899


/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
[I 2025-03-09 05:51:49,229] Trial 5 finished with value: 0.5014514895482254 and parameters: {'max_depth': 7, 'learning_rate': 0.04356992752665128, 'n_estimators': 264}. Best is trial 0 with value: 0.5885272665136787.
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 1.393902 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899


/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
[I 2025-03-09 05:53:59,657] Trial 6 finished with value: 0.027179320395438245 and parameters: {'max_depth': 6, 'learning_rate': 0.0003451272105417426, 'n_estimators': 261}. Best is trial 0 with value: 0.5885272665136787.
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 2.413440 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive ga

/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
[I 2025-03-09 05:58:04,824] Trial 7 finished with value: 0.5040089766892397 and parameters: {'max_depth': 5, 'learning_rate': 0.026122047313686676, 'n_estimators': 583}. Best is trial 0 with value: 0.5885272665136787.
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 4.092582 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive ga

/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
[I 2025-03-09 05:59:47,320] Trial 8 finished with value: 0.0009615290160892576 and parameters: {'max_depth': 2, 'learning_rate': 3.691854129724403e-05, 'n_estimators': 353}. Best is trial 0 with value: 0.5885272665136787.
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 1.972463 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899


/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
[I 2025-03-09 06:02:51,234] Trial 9 finished with value: 0.42321555995311533 and parameters: {'max_depth': 8, 'learning_rate': 0.020919238752723467, 'n_estimators': 257}. Best is trial 0 with value: 0.5885272665136787.


In [ ]:
study.best_params

{'max_depth': 15, 'learning_rate': 0.060611653131808045, 'n_estimators': 570}

In [ ]:
lgbm_model = LGBMRegressor(**study.best_params)
lgbm_model.fit(X_train, y_train)

train_preds = lgbm_model.predict(X_train)
test_preds = lgbm_model.predict(X_test)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(train_preds))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(test_preds))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), np.exp(train_preds))}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(test_preds))}")

/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


[LightGBM] [Info] Auto-choosing col-wise multi-threading, the overhead of testing was 2.164678 seconds.
You can set `force_col_wise=true` to remove the overhead.
[LightGBM] [Info] Total Bins 201960
[LightGBM] [Info] Number of data points in the train set: 69370, number of used features: 792
[LightGBM] [Info] Start training from score 5.626899


/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/utils/deprecation.py:151: FutureWarning: 'force_all_finite' was renamed to 'ensure_all_finite' in 1.6 and will be removed in 1.8.
  warnings.warn(


MAPE на трейне =1.4149554150557286
MAPE на тесте =6.244246928492248
r2 на трейне =0.39529258282978874
r2 на тесте =0.2445039235966332


In [13]:
from xgboost import XGBRegressor

boost = XGBRegressor(n_estimators=300)
boost.fit(X_train, y_train)

train_preds = boost.predict(X_train)
test_preds = boost.predict(X_test)

print(f"MAPE на трейне ={mean_absolute_percentage_error(np.exp(y_train), np.exp(train_preds))}")
print(f"MAPE на тесте ={mean_absolute_percentage_error(np.exp(y_test), np.exp(test_preds))}")

print(f"r2 на трейне ={r2_score(np.exp(y_train), np.exp(train_preds))}")
print(f"r2 на тесте ={r2_score(np.exp(y_test), np.exp(test_preds))}")

MAPE на трейне =0.42688412053257474
MAPE на тесте =7.061499139511124
r2 на трейне =0.7595334963430225
r2 на тесте =0.27625642890806357


###Результаты:
- DecisionTreeRegressor(max_depth=11)

MAPE на тесте =49.360855940656016

r2 на тесте =0.023259471421516076

- RandomForestRegressor(n_estimators=5, max_depth=11)

MAPE на тесте =44.718273501877356

r2 на тесте =-0.0057825433164413464

- XGBRegressor()

MAPE на тесте =2.298873468404665

r2 на тесте =0.2711146253672051

- CatBoostRegressor()

MAPE на тесте =4.315919766835618

r2 на тесте =0.33934997447637794

- LGBMRegressor(max_depth=16, n_estimators=623, learning_rate=0.10385498382901673)

MAPE на тесте =5.83013318938401

r2 на тесте =0.28789271804969485